### PR TITLE
feat(MA-1067): add API to check of teachers unattached to classes

### PR DESCRIPTION
**Pull request by flmartineau**
## Describe your changes

- add new API for checking teachers with no classes
- add new API for checking the initialization status of a structure

Related to https://github.com/OPEN-ENT-NG/presences/pull/296

API doc : https://confluence.support-ent.fr/display/MTVS/Init+API

## Checklist tests

- check the API routes are behaving properly: a response is received if the user has the proper rights.

## Issue ticket number and link

https://jira.support-ent.fr/browse/MA-1067

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

**Review by Rseuret — APPROVED** [review]

File: src/main/java/fr/openent/viescolaire/controller/InitController.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                    renderError(request);
+                })
+                .onSuccess(teachers -> renderJson(request, teachers.toJson()));
+    }
+
+
+    @Get("/structures/:structureId/initialization")
+    @ApiDoc("Retrieve initialization status")
+    @SecuredAction(value="", type = ActionType.RESOURCE)
+    @ResourceFilter(AdministratorRight.class)
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/controller/InitController.java:43]: doc d'api dans la description de la PR

File: src/main/java/fr/openent/viescolaire/controller/InitController.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    @Get("/structures/:structureId/initialization/teachers")
+    @ApiDoc("Retrieve teachers initialization status")
+    @SecuredAction(value="", type = ActionType.RESOURCE)
+    @ResourceFilter(AdministratorRight.class)
+    public void getTeachersInitializationStatus(HttpServerRequest request) {
+        String structureId = request.getParam(Field.STRUCTUREID);
+        this.initService.getTeachersStatus(structureId)
+                .onFailure(fail -> {
+                    log.error(String.format("[Viescolaire@%s::getTeachersInitializationStatus] Failed to retrieve " +
+                            "teachers initialization status", this.getClass().getSimpleName()), fail);
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/controller/InitController.java:33]: fail => fail.getMessage() ?

File: src/main/java/fr/openent/viescolaire/model/InitTeachers.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    private List<User> teachers;
+    private Integer count;
+
+    public InitTeachers() {
+        this.teachers = new ArrayList<>();
+        this.count = 0;
+    }
+
+    public InitTeachers(JsonObject json) {
+        this.teachers = json.getJsonArray(Field.TEACHERS).stream().map(o -> new User((JsonObject) o)).collect(Collectors.toList());
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/model/InitTeachers.java:20]:
```suggestion
        this.teachers = json.getJsonArray(Field.TEACHERS).stream()
           .filter(JsonObject.class::isInstance)
           .map(JsonObject.class::cast)
           .map(User.class::new)
           .collect(Collectors.toList());
```

File: src/main/java/fr/openent/viescolaire/service/InitService.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package fr.openent.viescolaire.service;
+
+import fr.openent.viescolaire.model.*;
+import io.vertx.core.*;
+
+public interface InitService {
+
+    Future<InitTeachers> getTeachersStatus(String structureId);
+
+    Future<Boolean> getInitializationStatus(String structureId);
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/service/InitService.java:10]: Possible une Java doc

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultInitService.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        JsonObject params = new JsonObject()
+                .put(Field.STRUCTUREID, structureId);
+
+        neo4j.execute(query, params, Neo4jResult.validResultHandler(res -> {
+            if (res.isRight()) {
+                JsonArray teachers = res.right().getValue();
+
+                InitTeachers initTeachers = new InitTeachers();
+                initTeachers.setCount(teachers.size());
+                initTeachers.setTeachers(teachers.stream().map(o -> new User((JsonObject) o)).collect(Collectors.toList()));
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultInitService.java:46]:
```suggestion
                initTeachers.setTeachers(teachers.stream()
                .filter(JSonObject.class::isInstance)
                .map(JsonObject.class::cast)
                .map(User.class::new)
                .collect(Collectors.toList()));
```

**Review by BastienBenard — COMMENTED** [review]
Est ce que tu peux faire des tests sur le nouveau service ?

File: src/main/java/fr/openent/viescolaire/controller/InitController.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    @Get("/structures/:structureId/initialization")
+    @ApiDoc("Retrieve initialization status")
+    @SecuredAction(value="", type = ActionType.RESOURCE)
+    @ResourceFilter(AdministratorRight.class)
+    public void getInitializationStatus(HttpServerRequest request) {
+        String structureId = request.getParam(Field.STRUCTUREID);
+        this.initService.getInitializationStatus(structureId)
+                .onFailure(fail -> {
+                    log.error(String.format("[Viescolaire@%s::getInitializationStatus] Failed to retrieve " +
+                            "initialization status", this.getClass().getSimpleName()), fail);
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/controller/InitController.java:49]: fail.getMessage() ?

**Review by JinDouang — CHANGES_REQUESTED** [review]

**Pull request merged**